engines/agos/drivers/accolade/adlib.cpp

```diff
@@ -286,19 +286,16 @@ void MidiDriver_Accolade_AdLib::setTimerCallback(void *timerParam, Common::Timer
 
 void MidiDriver_Accolade_AdLib::noteOn(byte FMvoiceChannel, byte note, byte velocity) {
 	byte adjustedNote     = note;
-	byte adjustedVelocity = velocity;
 	byte regValueA0h      = 0;
 	byte regValueB0h      = 0;
 
 	// adjust velocity
-	int16 channelVolumeAdjust = _channels[FMvoiceChannel].volumeAdjust;
-	channelVolumeAdjust += adjustedVelocity;
-	channelVolumeAdjust = CLIP<int16>(channelVolumeAdjust, 0, 0x7F);
-
+	int16 channelVolumeAdjust = velocity + _channels[FMvoiceChannel].volumeAdjust;
 	// adjust velocity with the master volume
-	byte volumeAdjust = adjustedVelocity * ((float) (128 + _masterVolume) / 128);
+	channelVolumeAdjust = (channelVolumeAdjust * (128 + _masterVolume)) / 128;
+	channelVolumeAdjust = CLIP<int16>(channelVolumeAdjust, 0, 0x7F);
 
-	adjustedVelocity = volumeAdjust;
+	byte adjustedVelocity = channelVolumeAdjust;
```

Comment:
Note that this value is later set on _channels[FMvoiceChannel].velocity. This is used in setVolume, where a new _masterVolume value is applied. So the value set there should be just velocity and volumeAdjust, without _masterVolume applied. Also, setVolume has the same bugged volume calculation.

Reply:
Ok, I'm getting pretty confused over here so I'm just going to write down some stuff for my own benefit here... :-)

Before the changes, we have the following variables:

- The **velocity** input parameter.
- **channelVolumeAdjust** which is calculated as CLIP<int16>(velocity + _channels[FMvoiceChannel].volumeAdjust)
- **volumeAdjust** which is calculated as velocity * ((floag) (128 + _masterVolume) / 128)
- **adjustedVelocity**, which is set to **volumeAdjust**.

After this point, **velocity**, **channelVolumeAdjust** and **volumeAdjust** are no longer used. (Which means that **channelVolumeAdjust** isn't used at all.)

**adjustedVelocity** is then further adjusted, before it's stored in _channels[FMvoiceChannel].velocity and passed on to noteOnSetVolume().

So if _masterVolume should not be factored into _channels[FMvoiceChannel].velocity, I guess it would be more sensible to do that adjustment in noteOnSetVolume()?

Reply:
I've made some further changes. Now the volume seems a bit low to me, but of course we could have a different value than 15 for the master volume.

Reply:
Yes, that will put it to half volume. I thought master volume was set from the user volume setting, but that does not seem to be the case. That should probably be added to MidiPlayer::setVolume (only for this driver).

Reply:
Just to confuse things - or at least me - even further, this is what AGOS's MidiPlayer::pause() does:

```
	if (musicType == MT_ADLIB && _musicMode == kMusicModeAccolade) {
		static_cast <MidiDriver_Accolade_AdLib*> (_driver)->setVolume(_paused ? 0 : 128);
	}
```

So here it seems like 0 means silence, and 128 means... unchanged volume? Half volume? Either way, 128 won't reset _masterVolume to 15.

I'm not sure how to proceed here...

Reply:
The range for the setVolume parameter is 0 - 255. This is then converted to -128 to 127 for _masterVolume, then back to 0 - 255 for the volume calculations because... reasons. The unpaused value should be the previously set volume, i.e. MidiPlayer::_musicVolume or _sfxVolume, whichever one you decide to use in MidiPlayer::setVolume.